haystack/nodes/prompt/invocation_layer/sagemaker.py

```diff
@@ -0,0 +1,221 @@
+    print(res)
+    """
+
+    def __init__(
+        self,
+        model_name_or_path: str,
+        max_length: int = 100,
+        aws_access_key_id: Optional[str] = None,
+        aws_secret_access_key: Optional[str] = None,
+        aws_session_token: Optional[str] = None,
+        region_name: Optional[str] = None,
+        profile_name: Optional[str] = None,
+        **kwargs,
+    ):
+        """
+        Instantiates the session with SageMaker using IAM based authentication via boto3.
+
+        :param model_name_or_path: The name for SageMaker Model Endpoint.
+        :param max_length: The maximum length of the output text.
+        :param aws_access_key_id: AWS access key ID.
+        :param aws_secret_access_key: AWS secret access key.
+        :param aws_session_token: AWS session token.
+        :param region_name: AWS region name.
+        :param profile_name: AWS profile name.
+        """
+        boto3_import.check()
+        super().__init__(model_name_or_path)
+        try:
+            session = boto3.Session(
+                aws_access_key_id=aws_access_key_id,
+                aws_secret_access_key=aws_secret_access_key,
+                aws_session_token=aws_session_token,
+                region_name=region_name,
+                profile_name=profile_name,
+            )
+            self.client = session.client("sagemaker-runtime")
+        except Exception as e:
+            raise SageMakerInferenceError(
+                f"Could not connect to SageMaker Inference Endpoint {model_name_or_path}."
+                f"Make sure the Endpoint exists and AWS environment is configured. {e}"
+            )
+        self.max_length = max_length
+
+        # for a list of supported parameters (TODO: verify if all of them are supported in sagemaker)
+        self.model_input_kwargs = {
+            key: kwargs[key]
+            for key in [
+                "best_of",
+                "details",
+                "do_sample",
+                "max_new_tokens",
+                "max_time",
+                "model_max_length",
+                "num_return_sequences",
+                "repetition_penalty",
+                "return_full_text",
+                "seed",
+                "temperature",
+                "top_k",
+                "top_p",
+                "truncate",
+                "typical_p",
+                "watermark",
+            ]
+            if key in kwargs
+        }
+
+        # We pop the model_max_length from the model_input_kwargs as it is not sent to the model
+        # but used to truncate the prompt if needed
+        model_max_length = self.model_input_kwargs.pop("model_max_length", 1024)
+
+        # Truncate prompt if prompt tokens > model_max_length-max_length (max_lengt is the length of the generated text)
+        self.prompt_handler = DefaultPromptHandler(
+            model_name_or_path="gpt2", model_max_length=model_max_length, max_length=self.max_length or 100
```

Comment:
Why is `model_name_or_path` set to `"gpt2"` here?

Reply:
We can't deduce the tokenizer type from the model endpoint name, so we must default to some tokenizer to somehow estimate the prompt length in tokenized form. Our options are limited here, hence gpt2. Open to ideas here